### PR TITLE
Ethereum RPC API migration to a coroutine implementation

### PR DESCRIPTION
**Pull request by jpalvarezl**
Closes https://github.com/gnosis/svalinn-kotlin/issues/39 .

Changes proposed in this pull request:
- RxJava extraction with analogous coroutine implementation
- Retrofit is optional, suplied as default implementation

@gnosis/mobile-devs

Tests still need to be transalated to coroutines, there are 2 that have already been migrated as example.

### CODE REVIEW

File: ethereum-rpc-retrofit/src/test/java/pm/gnosis/ethereum/rpc/retrofit/RetrofitEthereumRpcConnectorTest.kt [hunk 1]
```diff
@@ -30,55 +29,55 @@ class RetrofitEthereumRpcConnectorTest {
...
-        given(api.block(MockUtils.any())).willReturn(expected)
+        given(api.block(MockUtils.any())).willThrow(expected)
 
         assertEquals(expected, connector.block(request))
         then(api).should().block(request)
         then(api).shouldHaveNoMoreInteractions()
     }
 
-    @Test
+    @Test(expected = NotImplementedError::class)
```
Comment (rmeissner) [inline ethereum-rpc-retrofit/src/test/java/pm/gnosis/ethereum/rpc/retrofit/RetrofitEthereumRpcConnectorTest.kt:54]: If you do it this way, does it continue execution after it throws?

I would expect that not even the `assertEquals` is executed.

File: ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                it.to,
+                it.cumulativeGasUsed,
+                it.gasUsed,
+                it.contractAddress,
+                it.logs.map { TransactionReceipt.Event(it.logIndex, it.data, it.topics) }
+            )
+        } ?: throw TransactionReceiptNotFound()
+
+
+    override suspend fun getTransactionByHash(transactionHash: String): TransactionData =
```
Comment (rmeissner) [inline ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt:67]: This feels weird, as we have the whole logic twice (the part that builds the JsonRpcRequest). That should only exist once, right?

so what I would do here is create a scope (and maybe allow configuring that), use the sync interface and wrap the call in the scope.

Or in the first run we say this is not something that needs be solved here and just provide the sync interface. 

OR we say ... coroutines is the way to go (and we want to use them properly with retrofit) so for now we do not provide anything else.

Comment (elgatovital) [reply ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt:67]: I think it's worth considering supporting just coroutines. This is also the approach we use in the app. Having rx will create additional maintenance and tech debt and we don't know for sure if it is ever gonna be needed. Supporting both at the beginning will slow us down. We still can add support for rx at some later point.

File: ethereum-rpc-retrofit/src/test/java/pm/gnosis/ethereum/rpc/retrofit/RetrofitEthereumRpcConnectorTest.kt (hunk 1 above)
Comment (jpalvarezl) [reply ethereum-rpc-retrofit/src/test/java/pm/gnosis/ethereum/rpc/retrofit/RetrofitEthereumRpcConnectorTest.kt:54]: It does not. So basically the assertion is implicit with the expected exception type. We basically miss the assertion as well of method calls, which I guess are also important. I could create an extension function that `try/catches/asserts`

File: ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt (hunk 2 above)
Comment (jpalvarezl) [reply ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt:67]: Actually, another argument in favour of having only coroutines is that there are already adapters implemented that will wrap everything in some form of rx stream: https://github.com/Kotlin/kotlinx.coroutines/tree/master/reactive/kotlinx-coroutines-rx2

Comment (rmeissner) [reply ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt:67]: If we all agree then lets do that :D

**Comment by jpalvarezl** [comment]
While developping https://github.com/gnosis/safe-android/issues/557 I also realised that it would be nice to make all the `EthRequest` subclasses into data classes for ease of testing.

**Comment by rmeissner** [comment]
Could you give the PR a little bit more descriptive title? :)

**Pull request merged**